### 의사결정나무
#### 1. 데이터 > '알고리즘' > 모델
##### 1) 데이터에 내재된 패턴을 변수의 조합으로, 나타내는 예측/분류 모델을 나무의 형태로 만드는 것  
##### 2) 데이터를 2개 혹은 그 이상의 부분집합으로 분할, 데이터가 '균일'해지도록 분할
##### 3) 균일의 의미? 예측(비슷한 수치를 갖고 있는 관측치의 모음), 분류(비슷한 범주를 갖고 있는 관측치의 모음)
##### 4) 이진분할 : 부분집합의 개수 == 끝 마디 노드(Terminal Node) 개수
##### 5) 의사결정나무 모형 : 변수(차원)의 개수와 상관없이 2차원 공간에 나무의 형태로 표현 가능 및 해석력 우수

#### 2. 예측나무 모델(Regression Tree)
##### 1) 보통 새로운 데이터의 Y값은 해당 데이터가 속한 부분집합의 관측치들의 평균으로 나눠서 구한다.
##### 2) CmI 의 I는 Indicator Funtion : 중괄호에 있는 것이 참(True) > 1, 거짓(False) > 0, 
##### > x1, x2가 Rm 지역에 있느냐에 따라 참, 거짓 리턴
##### > 미래의 데이터가 R3 지역에 존재한다면? f(x) = C3

#### 3. 예측나무 모델링 프로세스
##### 1) Sum((실제 Y값 - 모델예측 Y값)^2) 최소화 되려면 , 각 분할에 속해있는 y값들의 평균으로 예측했을 때 오류가 최소이다. > 최상의 분할
##### 2) 예측모델에서 분할변수(j), 분할점(s)은 어떻게 결정할까?
##### > 분할변수(j) X 분할점(s)의 경우의 수를 모두 수행
##### > 기존 데이터에 대한 설명 : 각 경우의 수 마다 Sum((실제 Y값 - 해당 분할의 Y값들의 평균)^2) 작을 때까지 반복하면서 분할
##### > 미래 데이터에 대한 예측 : 분할의 Y값들의 평균으로 예측

#### 4. 분류나무 모델링 프로세스
##### 1) 관측치 마다 반응변수 값 yi = 1,2,3,,,,, 즉 k개의 클래스 존재
##### 2) Pmk = P(확률), m(끝 노드), k(클래스) = P11(1번 끝노드에서 1클래스에 속해 있는 관측치의 비율)
##### > 끝노드 m으로 분류된 관측치는 k(m) 클래스로 분류
##### 3) 분류 모델에서의 비용함수(불순도 측정) : Misclassification rate(실제 범주 ~ 모델 범주가 얼마나 매칭?), Gini Index , Cross-entropy
##### 4) 예측모델에서 분할변수(j), 분할점(s)은 어떻게 결정할까?(분할법칙)
##### > 분할변수와 분할점은 목표변수의 분포를 가장 잘 구별해주는 쪽으로 정함 - 순수도(Purity), 불순도(impurity)
##### > '불순도 감소'가 최대 혹은 '낮은비용' 함수 선택
##### 5) Information Gain - Cross Entropy : 엔트로피는 무질서도, 혼잡도 > 무질서, 혼잡소를 줄였다면 그 변수는 중요한 변수

#### 5. 개별 트리 모델의 단점 > 랜덤포레스트
##### 1) 계층적구조, 에러 전파
##### 2) 학습 데이터의 미세한 변동 > 최종 결과에 영향
##### 3) 적은 개수의 노이즈에도 크게 영향
##### 4) 최종노드 개수가 늘어나면 과적합 

## 1. boston데이터를 가지고 집값을 예측하는 의사결정나무 모델을 만들고 결과를 해석하세요 (예측)

In [2]:
import pandas as pd 
from sklearn.datasets import load_boston
boston= load_boston()
X = pd.DataFrame(boston.data,columns=boston.feature_names)
y = boston.target

C:\Users\KDB\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing dataset (i.

In [3]:
X

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       17.8  392.83   4.03  
3       18.7  394.63   2.94  
4       18.7  396.90   5.33  
..       ...     ...    ...  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88  

[506 rows x 13 columns]

In [4]:
y

array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
       18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
       15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
       13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
       21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
       35.4, 24.7, 31.6, 23.3, 19.6, 18.7, 16. , 22.2, 25. , 33. , 23.5,
       19.4, 22. , 17.4, 20.9, 24.2, 21.7, 22.8, 23.4, 24.1, 21.4, 20. ,
       20.8, 21.2, 20.3, 28. , 23.9, 24.8, 22.9, 23.9, 26.6, 22.5, 22.2,
       23.6, 28.7, 22.6, 22. , 22.9, 25. , 20.6, 28.4, 21.4, 38.7, 43.8,
       33.2, 27.5, 26.5, 18.6, 19.3, 20.1, 19.5, 19.5, 20.4, 19.8, 19.4,
       21.7, 22.8, 18.8, 18.7, 18.5, 18.3, 21.2, 19.2, 20.4, 19.3, 22. ,
       20.3, 20.5, 17.3, 18.8, 21.4, 15.7, 16.2, 18. , 14.3, 19.2, 19.6,
       23. , 18.4, 15.6, 18.1, 17.4, 17.1, 13.3, 17.8, 14. , 14.4, 13.4,
       15.6, 11.8, 13.8, 15.6, 14.6, 17.8, 15.4, 21

In [9]:
# 1. 데이터 전처리
# 1-1) 결측치 확인
X.info() # 결측치 X
# 1-2) 데이터 분할
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.3, random_state = 1000)
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
dtypes: float64(13)
memory usage: 51.5 KB
(354, 13)
(354,)
(152, 13)
(152,)


In [21]:
# 2. 데이터 적합
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error
import numpy as np

decisionTreeRegressor = DecisionTreeRegressor()
param_grid = {'max_depth' : range(1,40,2), 'min_samples_split': range(2,40,2)}
model = GridSearchCV(decisionTreeRegressor,param_grid)
model.fit(X_train,y_train)
y_pred = model.predict(X_test)
print('model_best_estimator : ',model.best_estimator_)
print('model_train_score : ', model.score(X_train, y_train)) 
print('model_test_score : ', model.score(X_test, y_test))
print('model_RMSE : ', np.sqrt(mean_squared_error(y_test, y_pred)))

model_best_estimator :  DecisionTreeRegressor(max_depth=33, min_samples_split=12)
model_train_score :  0.969129782677371
model_test_score :  0.6066843272221676
model_RMSE :  5.463759487314181


##### GridSearchCV로 DecisionTree 회귀분석의 최적화된 파라미터는 트리의 최대 깊이(max_depth)  33, 내부 노드를 분할하는데 필요한 최소 샘플 수(min_samples_split)  12이다. 
##### 최적의 파라미터로 생성된 모델의 Train Data의 설명력은 96%, Test Data의 설명력은 60% 이므로 과적합임을 알 수있다.
##### RMSE 값을 통해 실제값과 예측값은 평균 5.46 차이가 난다고 볼 수 있다.

## titanic데이터를 가지고 생존율을 예측하는 의사결정나무 모델을 만들고 결과를 해석하세요 (분류)

In [39]:
df_t = pd.read_csv("data/titanic.csv")
df_t

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [40]:
# 1. 데이터 전처리
# 1-1) 결측치 확인
df_t.info() # age, embarked(embark_town)

#1-2) 결측치 처리 전 중복 컬럼 제거(deck는 왜 제거 할까?)
df_t.drop(columns = ['class', 'embark_town','alive','adult_male','deck'], inplace = True)

# 1-3) 결측치 처리(평균대치법 혹은 최대값대치)
df_t['age'].fillna(df_t['age'].mean(), inplace = True)
df_t['embarked'].fillna(df_t['embarked'].value_counts().index[0], inplace = True)

df_t.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
 #  

In [41]:
# 1-4) 범주형 변수 처리
df_t # sex, embarked, who
df_t = pd.get_dummies(df_t, columns = ['sex','embarked','who'])
df_t 

survived  pclass        age  sibsp  parch     fare  alone  sex_female  \
0           0       3  22.000000      1      0   7.2500  False           0   
1           1       1  38.000000      1      0  71.2833  False           1   
2           1       3  26.000000      0      0   7.9250   True           1   
3           1       1  35.000000      1      0  53.1000  False           1   
4           0       3  35.000000      0      0   8.0500   True           0   
..        ...     ...        ...    ...    ...      ...    ...         ...   
886         0       2  27.000000      0      0  13.0000   True           0   
887         1       1  19.000000      0      0  30.0000   True           1   
888         0       3  29.699118      1      2  23.4500  False           1   
889         1       1  26.000000      0      0  30.0000   True           0   
890         0       3  32.000000      0      0   7.7500   True           0   

     sex_male  embarked_C  embarked_Q  embarked_S  who_child  who_man  \
0           1           0           0           1          0        1   
1           0           1           0           0          0        0   
2           0           0           0           1          0        0   
3           0           0           0           1          0        0   
4           1           0           0           1          0        1   
..        ...         ...         ...         ...        ...      ...   
886         1           0           0           1          0        1   
887         0           0           0           1          0        0   
888         0           0           0           1          0        0   
889         1           1           0           0          0        1   
890         1           0           1           0          0        1   

     who_woman  
0            0  
1            1  
2            1  
3            1  
4            0  
..         ...  
886          0  
887          1  
888          1  
889          0  
890          0  

[891 rows x 15 columns]

In [45]:
# 1-5) 데이터 분할
X = df_t.drop('survived', axis = 1)
y = df_t['survived']

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 2022)
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(623, 14)
(623,)
(268, 14)
(268,)


In [ ]:
# 2. DecisionTree Classifier 적합
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error
from numpy as np

decisionTreeRegressor = DecisionTreeClassifier()
param_grid
